File: algorithm/234. Palindrome Linked List/234.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* rev=NULL;
        ListNode* fast=head;
        ListNode* slow=head;
        ListNode* tem;
        ListNode* tem1;
        while(fast && fast->next)
        {
            fast=fast->next->next;
            tem=rev;
            tem1=slow->next;
            rev=slow;
            rev->next=tem;
            slow=tem1;
        }
        if(fast)
            slow=slow->next;
        while ( rev && rev->val==slow->val)
        {
            rev=rev->next;
            slow=slow->next;
        }
        return rev==NULL;
    }
};

File: algorithm/234. Palindrome Linked List/234.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        fast=slow=head
        rev=None
        while fast and fast.next:
            fast=fast.next.next
            slow,rev,rev.next=slow.next,slow,rev
        if fast:
            slow=slow.next
        while slow and slow.val==rev.val:
            slow=slow.next
            rev=rev.next
        return rev==None
